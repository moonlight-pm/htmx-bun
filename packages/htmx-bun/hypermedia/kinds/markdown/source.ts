import Markdown from "markdown-it";
import { Source } from "~/hypermedia";

export class MarkdownSource extends Source {
    readonly kind = "markdown";

    #html!: string;

    get attributes() {
        return "{}";
    }

    get template() {
        return JSON.stringify(this.#html);
    }

    compile() {
        this.#html = markdown.render(this.text);
    }
}

const markdown = new Markdown();

markdown.set({
    xhtmlOut: false,
});

markdown.renderer.rules.text = (tokens, idx) => {
    return tokens[idx].content;
};
markdown.disable("entity");
