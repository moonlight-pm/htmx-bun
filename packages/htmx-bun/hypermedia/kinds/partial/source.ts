import * as ts from "typescript";
import { AttributeTypeString, AttributeTypes, Source } from "~/hypermedia";
import {
    HtmlFragment,
    htmlStartIndex,
    parseSource,
    printHtml,
    walkHtml,
} from "~/hypermedia/template";

/**
 * A `.part` source file composed of an upper code section (action) and
 * a subsequent html section (template).  Both are optional.
 */
export class PartialSource extends Source {
    readonly kind = "partial";

    #template!: HtmlFragment;
    #action!: ts.SourceFile;
    #attributes?: ts.InterfaceDeclaration;

    /**
     * Compile the partial source into TypeScript that can be loaded as a module by the loader plugin.
     * @returns The TypeScript source code for the partial.
     */
    compile() {
        this.#template = parseSource(this.text);
        this.#action = ts.createSourceFile(
            "",
            this.text.slice(0, htmlStartIndex(this.text)),
            ts.ScriptTarget.Latest,
            true,
        );
        this.transformExpressions();
        this.tranformAction();
    }

    get action() {
        return ts.createPrinter().printFile(this.#action);
    }

    get attributes() {
        const attributes: AttributeTypes = {};
        if (this.#attributes) {
            for (const member of this.#attributes.members) {
                if (ts.isPropertySignature(member) && member.type) {
                    attributes[member.name.getText()] =
                        member.type.getText() as AttributeTypeString;
                }
            }
        }
        return JSON.stringify(attributes);
    }

    get template() {
        return JSON.stringify(printHtml(this.#template));
    }

    private transformExpressions() {
        walkHtml(this.#template, (node, { visitEachChild }) => {
            if (node.type === "fragment") {
                visitEachChild(node);
                return;
            }
            if (node.type === "element") {
                for (const attr of node.attrs) {
                    for (const value of attr.value) {
                        if (value.type === "expression") {
                            value.content = this.transformExpression(
                                value.content,
                            );
                        }
                    }
                }
                visitEachChild(node);
                return;
            }
            if (node.type === "expression") {
                node.content = this.transformExpression(node.content);
                return;
            }
        });
    }

    /**
     * Re-scopes expression property accessors to an identifier which will be provided
     * at runtime.
     *
     * @param expression - The expression to transform.
     * @returns The transformed expression.
     */
    private transformExpression(expression: string) {
        const source = ts.createSourceFile(
            "",
            expression,
            ts.ScriptTarget.Latest,
            true,
        );

        const transformer: ts.TransformerFactory<ts.Node> = (context) => {
            return (root) => {
                const visit: ts.Visitor = (node) => {
                    if (ts.isIdentifier(node)) {
                        if (ts.isExpressionStatement(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                        if (ts.isPropertyAccessExpression(node.parent)) {
                            if (node.parent.expression === node) {
                                return context.factory.createPropertyAccessExpression(
                                    context.factory.createIdentifier("$scope"),
                                    node,
                                );
                            }
                        }
                        if (ts.isTypeOfExpression(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                        if (ts.isBinaryExpression(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                        if (ts.isConditionalExpression(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                    }
                    return ts.visitEachChild(node, visit, context);
                };
                return ts.visitNode(root, visit) as ts.Node;
            };
        };

        const result = ts.transform(source, [transformer]);
        const transformed = result.transformed[0] as ts.SourceFile;
        const printer = ts.createPrinter({ omitTrailingSemicolon: true });
        return printer.printNode(
            ts.EmitHint.Unspecified,
            transformed.statements[0],
            source,
        );
    }

    /**
     * Final transformation of the TypeScript source, figuring out what identifiers need to be
     * pulled into the local scope, that we can then pass to the interpolation functions.
     *
     * We're also extracting out the partials Attributes interface, if it has one, so we can use
     * that information at runtime.  We also stub the $context local that will be passed
     * to the function that encloses the partial code.
     */
    private tranformAction() {
        const locals: string[] = ["$context"];
        const transformer: ts.TransformerFactory<ts.SourceFile> = (context) => {
            return (root) => {
                const statements: ts.Statement[] = [];
                const visit: ts.Visitor = (node) => {
                    node = ts.visitEachChild(node, visit, context);
                    if (
                        ts.isImportSpecifier(node) &&
                        node.isTypeOnly === false
                    ) {
                        locals.push(node.name.text);
                    }
                    if (ts.isImportDeclaration(node)) {
                        statements.push(node);
                        return;
                    }
                    if (
                        ts.isInterfaceDeclaration(node) &&
                        node.name.text === "Attributes"
                    ) {
                        this.#attributes = node;
                        statements.push(node);
                        return;
                    }
                    if (
                        ts.isVariableDeclaration(node) &&
                        node.parent.parent.parent === root
                    ) {
                        locals.push(node.name.getText());
                    }
                    return node;
                };
                root = ts.visitNode(root, visit) as ts.SourceFile;
                root = context.factory.updateSourceFile(root, [
                    ...statements,
                    ts.factory.createFunctionDeclaration(
                        [
                            ts.factory.createToken(ts.SyntaxKind.ExportKeyword),
                            // ts.factory.createToken(
                            //     ts.SyntaxKind.DefaultKeyword,
                            // ),
                            ts.factory.createToken(ts.SyntaxKind.AsyncKeyword),
                        ],
                        undefined,
                        ts.factory.createIdentifier("action"),
                        undefined,
                        [
                            ts.factory.createParameterDeclaration(
                                undefined,
                                undefined,
                                ts.factory.createIdentifier("$context"),
                                undefined,
                                ts.factory.createTypeReferenceNode(
                                    ts.factory.createIdentifier("Context"),
                                    undefined,
                                ),
                                undefined,
                            ),
                        ],
                        undefined,
                        ts.factory.createBlock(
                            [
                                ...root.statements,
                                ts.factory.createReturnStatement(
                                    ts.factory.createObjectLiteralExpression(
                                        locals.map((name) =>
                                            ts.factory.createShorthandPropertyAssignment(
                                                ts.factory.createIdentifier(
                                                    name,
                                                ),
                                            ),
                                        ),
                                        false,
                                    ),
                                ),
                            ],
                            true,
                        ),
                    ),
                ]);
                return root;
            };
        };
        const result = ts.transform(this.#action, [transformer]);
        this.#action = result.transformed[0];
    }
}
