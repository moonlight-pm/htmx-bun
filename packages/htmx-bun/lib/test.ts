import { afterAll } from "bun:test";
import { mkdtempSync, rmdirSync } from "node:fs";
import { tmpdir } from "node:os";

export function makeTemporaryDirectory() {
    const dir = mkdtempSync(`${tmpdir()}/htmx-bun-`);
    afterAll(() => {
        rmdirSync(dir, { recursive: true });
    });
    return dir;
}
