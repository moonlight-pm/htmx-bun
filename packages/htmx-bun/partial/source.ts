import * as ts from "typescript";
import { HtmlFragment } from "./ast";
import { parsePartial } from "./parser";
import { printHtml } from "./printer";
import { htmlStartIndex } from "./scanner";
import { walkHtml } from "./transform";

/**
 * Represents an attribute for a partia/tag.
 */
export interface Attribute {
    name: string;
    type: string;
}

export type Attributes = Attribute[];

/**
 * The Source class is responsible for reading an htmx-bun partial `.part` file,
 * disentangling into script and html parts, and returning the source code that
 * will be resolved by the loader in `plugins/partial.ts`.
 */
export class Source {
    #text!: string;
    #html!: HtmlFragment;
    #code!: ts.SourceFile;
    #attributes?: ts.InterfaceDeclaration;

    /**
     * Creates a new instance of the Source class from a string.
     * @param path The path of the source.
     */
    constructor(text: string) {
        this.#text = text;
    }

    /**
     * Creates a new instance of the Source class from a path.
     * @param path The path of the source.
     */
    static async fromPath(path: string) {
        return new Source(await Bun.file(path).text());
    }

    /**
     * Compile the partial source into TypeScript that can be loaded as a module by the loader plugin.
     * @returns The TypeScript source code for the partial.
     */
    async compile() {
        const code: string[] = [
            this.#text.slice(0, htmlStartIndex(this.#text)),
        ];
        this.#html = parsePartial(this.#text);
        this.#code = ts.createSourceFile(
            "",
            code.join("\n"),
            ts.ScriptTarget.Latest,
            true,
        );
        this.transformExpressions();
        this.transformCode();
        return [
            `export const attributes = ${JSON.stringify(this.attributes)};`,
            `export const html = \`${this.html}\`;`,
            "",
            this.code,
        ].join("\n");
    }

    /**
     * Prints the internalathe serialized HTML
     * @returns The serialized HTML.
     */
    private get html() {
        return printHtml(this.#html);
    }

    /**
     * Prints the internal TypeScript code.
     * @returns The TypeScript code.
     */
    private get code() {
        return ts.createPrinter().printFile(this.#code);
    }

    private get attributes() {
        const attributes: Attributes = [];
        if (this.#attributes) {
            for (const member of this.#attributes.members) {
                if (ts.isPropertySignature(member) && member.type) {
                    attributes.push({
                        name: member.name.getText(),
                        type: member.type.getText(),
                    });
                }
            }
        }
        return attributes;
    }

    private transformExpressions() {
        walkHtml(this.#html, (node, { visitEachChild }) => {
            if (node.type === "fragment") {
                visitEachChild(node);
                return;
            }
            if (node.type === "element") {
                for (const attr of node.attrs) {
                    for (const value of attr.value) {
                        if (value.type === "expression") {
                            value.content = this.transformExpression(
                                value.content,
                            );
                        }
                    }
                }
                visitEachChild(node);
                return;
            }
            if (node.type === "expression") {
                node.content = this.transformExpression(node.content);
                return;
            }
        });
    }

    /**
     * Re-scopes expression property accessors to an identifier which will be provided
     * at runtime.
     *
     * @param expression - The expression to transform.
     * @returns The transformed expression.
     */
    private transformExpression(expression: string) {
        const source = ts.createSourceFile(
            "",
            expression,
            ts.ScriptTarget.Latest,
            true,
        );

        const transformer: ts.TransformerFactory<ts.Node> = (context) => {
            return (root) => {
                const visit: ts.Visitor = (node) => {
                    if (ts.isIdentifier(node)) {
                        if (ts.isExpressionStatement(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                        if (ts.isPropertyAccessExpression(node.parent)) {
                            if (node.parent.expression === node) {
                                return context.factory.createPropertyAccessExpression(
                                    context.factory.createIdentifier("$scope"),
                                    node,
                                );
                            }
                        }
                        if (ts.isTypeOfExpression(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                        if (ts.isBinaryExpression(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("$scope"),
                                node,
                            );
                        }
                    }
                    return ts.visitEachChild(node, visit, context);
                };
                return ts.visitNode(root, visit) as ts.Node;
            };
        };

        const result = ts.transform(source, [transformer]);
        const transformed = result.transformed[0] as ts.SourceFile;
        const printer = ts.createPrinter({ omitTrailingSemicolon: true });
        return printer.printNode(
            ts.EmitHint.Unspecified,
            transformed.statements[0],
            source,
        );
    }

    /**
     * Final transformation of the TypeScript source, figuring out what identifiers need to be
     * pulled into the local scope, that we can then pass to the interpolation functions.
     *
     * We're also extracting out the partials Attributes interface, if it has one, so we can use
     * that information at runtime.  We also stub Attribute and Helper locals, these will be passed
     * to the function that encloses the partial code, provided to it's local namespace.
     * @private
     */
    private transformCode() {
        const locals: string[] = ["Helper", "Attributes"];
        const transformer: ts.TransformerFactory<ts.SourceFile> = (context) => {
            return (root) => {
                const statements: ts.Statement[] = [];
                const visit: ts.Visitor = (node) => {
                    node = ts.visitEachChild(node, visit, context);
                    if (
                        ts.isImportSpecifier(node) &&
                        node.isTypeOnly === false
                    ) {
                        locals.push(node.name.text);
                    }
                    if (ts.isImportDeclaration(node)) {
                        statements.push(node);
                        return;
                    }
                    if (
                        ts.isInterfaceDeclaration(node) &&
                        node.name.text === "Attributes"
                    ) {
                        this.#attributes = node;
                        statements.push(node);
                        return;
                    }
                    if (
                        ts.isVariableDeclaration(node) &&
                        node.parent.parent.parent === root
                    ) {
                        locals.push(node.name.getText());
                    }
                    return node;
                };
                root = ts.visitNode(root, visit) as ts.SourceFile;
                root = context.factory.updateSourceFile(root, [
                    ...statements,
                    ts.factory.createFunctionDeclaration(
                        [
                            ts.factory.createToken(ts.SyntaxKind.ExportKeyword),
                            ts.factory.createToken(
                                ts.SyntaxKind.DefaultKeyword,
                            ),
                            ts.factory.createToken(ts.SyntaxKind.AsyncKeyword),
                        ],
                        undefined,
                        undefined,
                        undefined,

                        [
                            ts.factory.createParameterDeclaration(
                                undefined,
                                undefined,
                                ts.factory.createIdentifier("Helper"),
                                undefined,
                                ts.factory.createTypeReferenceNode(
                                    ts.factory.createIdentifier("Helper"),
                                    undefined,
                                ),
                                undefined,
                            ),
                            ts.factory.createParameterDeclaration(
                                undefined,
                                undefined,
                                ts.factory.createIdentifier("Attributes"),
                                undefined,
                                ts.factory.createTypeReferenceNode(
                                    ts.factory.createIdentifier("Attributes"),
                                    undefined,
                                ),
                                undefined,
                            ),
                        ],
                        undefined,
                        ts.factory.createBlock(
                            [
                                ...root.statements,
                                ts.factory.createReturnStatement(
                                    ts.factory.createObjectLiteralExpression(
                                        locals.map((name) =>
                                            ts.factory.createShorthandPropertyAssignment(
                                                ts.factory.createIdentifier(
                                                    name,
                                                ),
                                            ),
                                        ),
                                        false,
                                    ),
                                ),
                            ],
                            true,
                        ),
                    ),
                ]);
                return root;
            };
        };
        const result = ts.transform(this.#code, [transformer]);
        this.#code = result.transformed[0];
    }
}
