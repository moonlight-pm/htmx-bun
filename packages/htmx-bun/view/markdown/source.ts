import markdown from "markdown-it";

export class MarkdownSource {
    #text: string;

    /**
     * Creates a new instance of the MarkdownSource class from a string.
     * @param path The path of the source.
     */
    constructor(text: string) {
        this.#text = text;
    }

    /**
     * Creates a new instance of the MarkdownSource class from a path.
     * @param path The path of the source.
     */
    static async fromPath(path: string) {
        return new MarkdownSource(await Bun.file(path).text());
    }

    /**
     * Compile the source into TypeScript that can be loaded as a module by the loader plugin.
     * @returns The TypeScript source code.
     */
    async compile() {
        const html = markdown().render(this.#text);
        return `export const html = ${JSON.stringify(html)};`;
    }
}
