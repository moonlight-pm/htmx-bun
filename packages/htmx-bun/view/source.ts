import * as ts from "typescript";
import { matchRecursive } from "xregexp";
import { formatTypeScript } from "~/lib/format";
import {
    HtmlFragment,
    HtmlText,
    createHtmlFragment,
    parseHtml,
    printHtmlSyntaxTree,
    transformHtml,
} from "~/lib/html";

export class Source {
    #html!: HtmlFragment;
    #code!: ts.SourceFile;
    #attributes?: ts.InterfaceDeclaration;
    #expressionIndex = 0;

    constructor(public path: string) {}

    async compile() {
        const text = await Bun.file(this.path).text();
        await this.disentangle(text);
        this.transformCode();
        // console.log(await this.code());
        // console.log(await this.html());
        return await formatTypeScript(`
            ${await this.code()}
            export const html = ${JSON.stringify(await this.html())};
            export const code = ${JSON.stringify(await this.code())};
            export const attributes = ${JSON.stringify(this.attributes)};
        `);
    }

    async html() {
        return await printHtmlSyntaxTree(this.#html);
    }

    async code() {
        return ts.createPrinter().printFile(this.#code);
    }

    get attributes() {
        const attributes: Attributes = [];
        if (this.#attributes) {
            for (const member of this.#attributes.members) {
                if (ts.isPropertySignature(member) && member.type) {
                    attributes.push({
                        name: member.name.getText(),
                        type: member.type.getText(),
                    });
                }
            }
        }
        return attributes;
    }

    private async disentangle(text: string) {
        this.#html = parseHtml(text);
        const code: string[] = [];
        await transformHtml(this.#html, async (node, { visitEachChild }) => {
            if (node.type === "element") {
                if (node.tag === "server") {
                    code.push((node.children[0] as HtmlText).content);
                    return createHtmlFragment(node.parent);
                }
                for (const attr of node.attrs) {
                    attr.value = this.matchExpressions(code, attr.value);
                }
            } else if (node.type === "text") {
                node.content = this.matchExpressions(code, node.content);
            }
            await visitEachChild(node);
            return node;
        });
        this.#code = ts.createSourceFile(
            "",
            code.join("\n"),
            ts.ScriptTarget.Latest,
            true,
        );
    }

    private matchExpressions(code: string[], text: string) {
        for (const expression of matchRecursive(text, "\\{", "\\}", "g")) {
            const $exp = `$exp${this.#expressionIndex++}`;
            code.push(
                `const ${$exp} = (env) => (${this.transformExpression(
                    expression,
                )});`,
            );
            text = text.replace(`{${expression}}`, $exp);
        }
        return text;
    }

    private transformExpression(expression: string) {
        const source = ts.createSourceFile(
            "",
            expression,
            ts.ScriptTarget.Latest,
            true,
        );

        const transformer: ts.TransformerFactory<ts.Node> = (context) => {
            return (root) => {
                const visit: ts.Visitor = (node) => {
                    if (ts.isIdentifier(node)) {
                        if (ts.isExpressionStatement(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("env"),
                                node,
                            );
                        }
                        if (ts.isPropertyAccessExpression(node.parent)) {
                            if (node.parent.expression === node) {
                                return context.factory.createPropertyAccessExpression(
                                    context.factory.createIdentifier("env"),
                                    node,
                                );
                            }
                        }
                        if (ts.isTypeOfExpression(node.parent)) {
                            return context.factory.createPropertyAccessExpression(
                                context.factory.createIdentifier("env"),
                                node,
                            );
                        }
                    }
                    return ts.visitEachChild(node, visit, context);
                };
                return ts.visitNode(root, visit) as ts.Node;
            };
        };

        const result = ts.transform(source, [transformer]);
        const transformed = result.transformed[0] as ts.SourceFile;
        const printer = ts.createPrinter({ omitTrailingSemicolon: true });
        return printer.printNode(
            ts.EmitHint.Unspecified,
            transformed.statements[0],
            source,
        );
    }

    private transformCode() {
        const locals: string[] = ["Helper", "Attributes"];
        const transformer: ts.TransformerFactory<ts.SourceFile> = (context) => {
            return (root) => {
                const statements: ts.Statement[] = [];
                const visit: ts.Visitor = (node) => {
                    node = ts.visitEachChild(node, visit, context);
                    if (
                        ts.isImportSpecifier(node) &&
                        node.isTypeOnly === false
                    ) {
                        locals.push(node.name.text);
                    }
                    if (ts.isImportDeclaration(node)) {
                        statements.push(node);
                        return;
                    }
                    if (
                        ts.isInterfaceDeclaration(node) &&
                        node.name.text === "Attributes"
                    ) {
                        this.#attributes = node;
                        statements.push(node);
                        return;
                    }
                    if (
                        ts.isVariableDeclaration(node) &&
                        node.parent.parent.parent === root
                    ) {
                        locals.push(node.name.getText());
                    }
                    return node;
                };
                root = ts.visitNode(root, visit) as ts.SourceFile;
                root = context.factory.updateSourceFile(root, [
                    ...statements,
                    ts.factory.createFunctionDeclaration(
                        [
                            ts.factory.createToken(ts.SyntaxKind.ExportKeyword),
                            ts.factory.createToken(ts.SyntaxKind.AsyncKeyword),
                        ],
                        undefined,
                        ts.factory.createIdentifier("$run"),
                        undefined,

                        [
                            ts.factory.createParameterDeclaration(
                                undefined,
                                undefined,
                                ts.factory.createIdentifier("Helper"),
                                undefined,
                                ts.factory.createTypeReferenceNode(
                                    ts.factory.createIdentifier("Helper"),
                                    undefined,
                                ),
                                undefined,
                            ),
                            ts.factory.createParameterDeclaration(
                                undefined,
                                undefined,
                                ts.factory.createIdentifier("Attributes"),
                                undefined,
                                ts.factory.createTypeReferenceNode(
                                    ts.factory.createIdentifier("Attributes"),
                                    undefined,
                                ),
                                undefined,
                            ),
                        ],
                        undefined,
                        ts.factory.createBlock(
                            [
                                ...root.statements,
                                ts.factory.createReturnStatement(
                                    ts.factory.createObjectLiteralExpression(
                                        locals.map((name) =>
                                            ts.factory.createShorthandPropertyAssignment(
                                                ts.factory.createIdentifier(
                                                    name,
                                                ),
                                            ),
                                        ),
                                        false,
                                    ),
                                ),
                            ],
                            true,
                        ),
                    ),
                ]);
                return root;
            };
        };
        const result = ts.transform(this.#code, [transformer]);
        this.#code = result.transformed[0];
    }
}

export interface Attribute {
    name: string;
    type: string;
}

export type Attributes = Attribute[];
