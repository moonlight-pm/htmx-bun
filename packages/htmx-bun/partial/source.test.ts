import { expect, test } from "bun:test";
import { Source } from "./source";

const source1 = `
interface Attributes {
    id: number;
    name: string;
}

const items = [{ id: 1, name: "one" }, { id: 2, name: "two" }];

<div>
    {id}
</div>
<div>{name}</div>
<ul>
    <li mx-each={items} mx-as="item" id={item.id}>{item.name}</li>
</ul>
`;
test("compile", async () => {
    const source = new Source(source1);
    const target = await source.compile();
    // console.log(target);
    expect(target).toMatchSnapshot();
    // expect(await source.code()).toMatchSnapshot();
});
