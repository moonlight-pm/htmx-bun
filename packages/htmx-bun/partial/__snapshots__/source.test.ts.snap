// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`compile 1`] = `
"export const attributes = [{"name":"id","type":"number"},{"name":"name","type":"string"}];
export const html = \`<div>
    {$scope.id}
</div>
<div>{$scope.name}</div>
<ul>
    <li mx-each={$scope.items} mx-as="item" id={$scope.item.id}>{$scope.item.name}</li>
</ul>
\`;

interface Attributes {
    id: number;
    name: string;
}
export default async function (Helper: Helper, Attributes: Attributes) {
    const items = [{ id: 1, name: "one" }, { id: 2, name: "two" }];
    return { Helper, Attributes, items };
}
"
`;
